src/clients/InventoryClient.ts

```diff
@@ -1105,6 +1106,183 @@ export class InventoryClient {
+
+    await sdkUtils.forEachAsync(this.getL1Tokens(), async (l1Token) => {
+      // We do not currently count any outstanding L2->L1 pending withdrawal balance in the cumulative balance
+      // because it can take so long for these withdrawals to finalize (usually >1 day and up to 7 days). Unlike the
+      // L1->L2 pending deposit balances which will finalize in <1 hour in most cases. For allocation % calculations,
+      // these pending withdrawals are therefore ignored.
+      const cumulativeBalance = this.getCumulativeBalance(l1Token);
+      if (cumulativeBalance.eq(bnZero)) {
+        return;
+      }
+      await sdkUtils.forEachAsync(chainIds, async (chainId) => {
+        if (chainId === this.hubPoolClient.chainId || !this._l1TokenEnabledForChain(l1Token, chainId)) {
+          return;
+        }
+
+        const l2Tokens = this.getRemoteTokensForL1Token(l1Token, chainId);
+        await sdkUtils.forEachAsync(l2Tokens, async (l2Token) => {
+          const tokenConfig = this.getTokenConfig(l1Token, chainId, l2Token);
+          if (!isDefined(tokenConfig)) {
+            return;
+          }
+          // When l2 token balance exceeds threshold, withdraw (balance - target) to hub pool.
+          const { targetOverageBuffer, targetPct, withdrawExcessPeriod } = tokenConfig;
+
+          // Excess withdrawals are activated only for chains where the withdrawExcessPeriod variable is set.
+          if (!isDefined(withdrawExcessPeriod)) {
+            return;
+          }
+
+          const adapter = this.adapterManager.adapters[chainId];
+          if (!adapter.isSupportedL2Bridge(l1Token)) {
+            this.logger.warn({
+              at: "InventoryClient#withdrawExcessBalances",
+              message: `No L2 bridge configured for ${getNetworkName(chainId)} for token ${l1Token}`,
+            });
+            return;
+          }
+
+          const currentBalance = this.tokenClient.getBalance(chainId, l2Token);
+          const currentAllocPct = this.getCurrentAllocationPct(l1Token, chainId, l2Token);
+
+          // We apply a discount on the effective target % because the repayment chain choice
+          // algorithm should never allow the inventory to get above the target pct * target overage buffer.
+          // Withdraw excess when current allocation % is within a small % of the target percentage multiplied
+          // by the target overage buffer.
+          const discountToTargetOverageBuffer = toBNWei("0.95");
+          const targetPctMultiplier = targetOverageBuffer.mul(discountToTargetOverageBuffer).div(this.scalar);
+          assert(
+            targetPctMultiplier.gte(toBNWei("1")),
+            `Target overage buffer multiplied by discount must be >= 1, got ${targetPctMultiplier.toString()}`
+          );
+          const excessWithdrawThresholdPct = targetPct.mul(targetPctMultiplier).div(this.scalar);
+
+          const l1TokenInfo = getL1TokenInfo(l2Token, Number(chainId));
```

Comment:
Minor: `chainId` is already a number.

```suggestion
          const l1TokenInfo = getL1TokenInfo(l2Token, chainId);
```